docs/type_inference.md

```diff
@@ -26,12 +26,21 @@ Both
 - **verifiers**: implemented by `Op::verify()`, and
 - **shape functions**: implemented by `InferTypeOpInterface` like `Op::inferReturnTypes()` or `Op::inferReturnTypeComponents`
 
-may have verification code to check operands/attributes/results. An ideal split would be that: let the verifiers check the operands/attributes, then let shape functions to calculate inferred result types and check the compatibility against the real result types. However, in reality this split has a few problems:
+may have verification code to check operands/attributes/results. An initial split would be that: let the verifiers check the operands/attributes, then let shape functions only calculate inferred result types and check the compatibility against the real result types. However, in reality this split has a few problems:
 
-1. Duplicated code: for example in verifiers we do some processing on the operands then verify some intermediate results, then in shape functions these intermediate results are useful to infer the final results. These intermediate results have to be calculated twice.
-2. Maintenance burden: as verifications of an op are contained in two different methods.
+- The shape function can be called by the auto-gen `builder()` alone, without calling the verifier first. So the related inputs must be verified in shape function as well.
+- Duplicated code: for example in verifiers we do some processing on the operands then verify some intermediate results, then in shape functions these intermediate results are useful to infer the final results. These intermediate results have to be calculated twice.
+- Maintenance burden: as verifications of an op are contained in two different methods.
+
+The solution is as follows:
+
+1. **For most ops w/o regions** (like PadOp):
```

Comment:
"without regions". Let's expand the abbreviation here since it doesn't save that much typing.